[#1348] Add chunk system for tile maps with procedural content generation support

jurca opened the pull request:
===:construction_worker_man: TODO checklist :construction_worker_man:===

- [x] add unit tests
- [x] integrate with `Scene`
- [x] integrate with `Engine`
- [x] integrate with physics
- [x] add a demo (chunk generator & garbage collector)
- [x] update the changelog
- [x] API documentation

===:clipboard: PR Checklist :clipboard:===

- [x] :pushpin: issue exists in github for these changes
- [x] :microscope: existing tests still pass
- [x] :see_no_evil: code conforms to the [style guide](https://github.com/excaliburjs/Excalibur/blob/master/STYLEGUIDE.md)
- [x] :triangular_ruler: new tests written and passing / old tests updated with new scenario(s)
- [x] :page_facing_up: changelog entry added (or not needed)

==================

Closes #1348

## Changes:

- Add the `ChunkSystemTileMap` and `ChunkSystemTileMapImpl` classes and the `ChunkSystemTileMapCollisionDetection` actor trait

jurca commented:
I feel at this point confident enough about the implementation and test coverage.

Would someone be, please, willing to review this PR and provide me with feedback before I'll add API documentation? I would prefer to add documentation afterwards in case the API is modified so that any inconsistencies are avoided.

eonarheim commented:
@jurca I'll have some time today to review 👍

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+  chunkSystemTileMap: ChunkSystemTileMap,
+  engine: Engine
+) => TileMap;
+export type SimpleCellGenerator = (
```
I did not feel comfortable to tie the cell generator wrapper's implementation to how the `TileMap` organizes its cells in the `data` property as it felt like closely entangling two separate implementation details.

But I will defer to your judgement here.

I have renamed the type and changed the generator's behavior.

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+import { Cell, TileMap } from './TileMap';
+
+export type ChunkGenerator = (chunkColumn: number, chunkRow: number, chunkSystemTileMap: ChunkSystemTileMap, engine: Engine) => TileMap;
+export type SimpleChunkGenerator = (
```
Correct, every chunk is represented by a `TileMap`. I have renamed the type.

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+  return Math.floor(number) === number && Math.abs(number) <= 9007199254740991;
+}
+
+function assign<T extends object, E extends object>(target: T, extension: E): T & E {
```
Done :)

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+  return extendedTarget;
+}
+
+function objectEntries<T>(object: { [s: string]: T } | ArrayLike<T>): [string, T][] {
```
Done :)

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+  });
+}
+
+function isSafeInteger(number: number): boolean {
```
I have updated the Polyfill.ts and lib settings in tsconfig.json. I had to include more polyfills that originally intended since [some typescript lib definitions](https://github.com/microsoft/TypeScript/blob/master/lib/lib.es2015.core.d.ts) are not very finely granular.

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+    const { cols, rows } = chunk;
+    for (let row = 0; row < rows; row++) {
+      for (let column = 0; column < cols; column++) {
+        simpleGenerator(chunk.getCell(column, row), chunkCellColumn + column, chunkCellRow + row, chunk, chunkSystemTileMap, engine);
```
I have updated the generator's behavior.

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+ */
+export class ChunkSystemTileMap extends Configurable(ChunkSystemTileMapImpl) {}
+
+export function wrapSimpleChunkGenerator(simpleGenerator: SimpleChunkGenerator): ChunkGenerator {
```
I have renamed the function and the parameter.

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+  };
+}
+
+export function wrapSimpleCellGenerator(simpleGenerator: SimpleCellGenerator): ChunkGenerator {
```
I have renamed the function and the parameter.

jurca commented on src/stories/ChunkSystemTileMap.stories.ts:
```diff
@@ -0,0 +1,146 @@
+    y: -(CHUNK_ROWS * CHUNK_SIZE * CHUNK_CELL_HEIGHT) / 2,
+    cols: CHUNK_COLS * CHUNK_SIZE,
+    rows: CHUNK_ROWS * CHUNK_SIZE,
+    chunkGenerator: wrapSimpleCellGenerator(cellGenerator),
```
Unfortunately, I do not see how to achieve that (without getting native support in JS for type reflection in browsers). The generator is always a function and we cannot rely on the number of function's parameters, since it does not have to use all of them.

However, if you have any ideas, I am open to suggestions.

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+        };
+        chunk.update(virtualEngine, delta);
+
+        chunk.renderingCache = document.createElement('canvas');
```
Are you referring to a possibility of a memory leak or ability of client code to access the canvas?

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+import { SpriteSheet } from './Drawing/SpriteSheet';
+import { Cell, TileMap } from './TileMap';
+
+export type ChunkGenerator = (chunkColumn: number, chunkRow: number, chunkSystemTileMap: ChunkSystemTileMap, engine: Engine) => TileMap;
```
While doable using this trick:

```typescript
type BaseChunkGenerator = (chunk: TileMap, ...args: Parameters<ChunkGenerator>) => TileMap
```

...I would argue against this, since numerical coordinates passed to `ChunkGenerator` and `BaseChunkGenerator` have different names and semantics (chunk-level coordinates vs cell-level coordinates).

That being said, you are the member of this project and I will follow by your lead. What is your stance on this?

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+      if (this._chunkRenderingCachePredicate && this._chunkRenderingCachePredicate(chunk, this, engine)) {
+        // We trick the TileMap chunk into assuming it is entirely visible on the screen, forcing it to render all its cells so that we may
+        // cache the rendering result.
+        const virtualEngine = Object.create(engine);
```
I've changed the code and introduced the `offScreenCulling` flag on `TileMap` to make the culling optional.

I have considered using a single virtual engine for a while, but as you have suggested, it's better to just skip the calculation completely.

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+  public readonly chunkCols: number;
+  public readonly chunkRows: number;
+  public readonly chunkGenerator: ChunkGenerator;
+  public readonly chunkGarbageCollectorPredicate: ChunkSystemGarbageCollectorPredicate;
```
It all comes down to the expected maximum size of the game world. For a "common" game, a `() => false` predicate is the best option since the world should fit into the memory. I am willing to make this the default implementation, if you agree.

However, for games aiming for a very large (or seemingly inifnite) world, it would be recommended to return `true` for any chunk that is "far enough" from the viewport. Such game would however need to handle the cases of actors trying to pathfind through garbage-collected chunks - it might be necessary for the predicate to communicate with other parts of the game's logic.

jurca replied:
After some thought, I've made the parameter optional, defaulting to no garbage collection, which is the best (safest) option for many games.

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+    }
+  }
+
+  private _updateChunk(chunkX: number, chunkY: number, engine: Engine, delta: number): CachedTileMap {
```
I've split the code into 3 methods.

jurca replied:
While we're on the topic, would you like me to split `_garbageCollectChunks` as well?

jurca commented on src/engine/Traits/ChunkSystemTileMapCollisionDetection.ts:
```diff
@@ -0,0 +1,34 @@
+import { CollisionType } from '../Collision/CollisionType';
+import { BoundingBox } from '../Collision/Index';
+
+export class ChunkSystemTileMapCollisionDetection implements Trait {
```
I have refactored both traits. Since the implementations were isomorphic, the original code has been removed.

jurca commented on src/engine/ChunkSystemTileMap.ts:
```diff
@@ -0,0 +1,466 @@
+    }
+  }
+
+  public collides(actor: Actor): Vector | null {
```
I have moved the ChunkSystemTileMap's implementation, and deprecated TileMap.collides in favor of the new shared code. Since there are minor cosmetic differences, I think it to be safest to preserve the current implementation of TileMap.collides for a while.

jurca commented:
@eonarheim I did my best to address your feedback (thanks a lot for it!). Could you, please, take another look once you can spare a moment?

eonarheim commented:
@jurca Thanks again for all your hard work! I'll have some time later today to review 👍

jurca commented:
@eonarheim My apologies if I'm being obnoxious or rude here, but did you manage to find some time to review this, please?

eonarheim commented:
@jurca Apologies, I'll get to this as soon as I can. It's been a busy couple weeks

jurca commented:
@eonarheim It's OK, I understand. And thanks a lot :).

jurca commented:
@eonarheim Apologies for making a change to the API at this point, I added an additional predicate that enables the game to garbage-collect pre-rendered chunk images without having to garbage-collect entire chunks. I've also updated the related documentation and added unit tests.

jurca commented:
@eonarheim Hello again, did you manage to find some time to review this?

jurca commented:
@eonarheim I hope you are all right. If you ever managed to take a look at this pull request, that would be great :heart: .

eonarheim commented:
Hi @jurca 

I want to apologize for the extreme delay. Thank you for your time and effort on this PR, we do really appreciate your work.

Unfortunately, we can't merge this PR in its current form. If you would still like to continue working on this feature, we'd like to discuss your approach further. I understand if you are no longer interested.

There is some code duplication and large changes in the TileMap API we aren't ready to make. It makes sense that there are large changes for such a big feature. In retrospect, we should have had a more in depth design discussion with you.

We believe you have the right approach; building new TileMaps for each chunk of the generation is a good idea. If you want, we can have a more detailed design discussion with you to integrate this feature into Excalibur as a series of smaller PRs. We can use this PR as a foundation for any future PRs.

Thanks again,
- Erik

jurca commented:
@eonarheim 

:disappointed: 

Alright, I understand. This happens. I suspected for a while that this PR would also make things more difficult for this one: https://github.com/excaliburjs/Excalibur/pull/1730, which might be quite undesirable.

I'll be happy to help and participate in said discussion, just a little forewarning: it might take me several days to respond every time throughout this year (but I'll do my best do reply quickly).
